AX: AccessibilityObject::visibleCharacterRange() returns the wrong result for vertical writing mode text that is only partially visible

twilco opened the pull request:
#### 9ac2c0188b7d0974774a40da80a628b1bd82b89e
<pre>
AX: AccessibilityObject::visibleCharacterRange() returns the wrong result for vertical writing mode text that is only partially visible
<a href="https://bugs.webkit.org/show_bug.cgi?id=283330">https://bugs.webkit.org/show_bug.cgi?id=283330</a>
<a href="https://rdar.apple.com/140158374">rdar://140158374</a>

Reviewed by Chris Fleizach.

AccessibilityObject::visibleCharacterRange() is implemented by shrinking the start and end of the element&apos;s rect
line-by-line until it fits within the viewport. The problem was that the logic for determining whether to start or
continue shrinking lines was based on checking rect.location() for shrinking the start of the text, or rect.location() + rect.size()
for the shrinking the end. This is wrong for &quot;flipped&quot; writing-modes (e.g. vertical-rl), whose `x` represents the end
of the text, not the start. This meant we would shrink the end lines when we actually should&apos;ve been shrinking the start,
and vice versa (hence this problem being specific to &quot;flipped&quot; writing modes).

With this commit, this logic now properly checks for and handles flipped writing modes. A test is added to prove we
behave correctly.

This commit also critically removes AccessibilityObject::m_cachedVisibleCharacterRange and AccessibilityObject::m_cachedVisibleCharacterRangeInputs.
These caches were added to make repeated calls for the same visibleCharacterRange fast, as Books seemed to be doing this
in the past. However, it doesn&apos;t seem to be doing this anymore. This cache was not a great idea in the first place (I added it).
It adds 120 bytes of size to every single AccessibilityObject, even if nothing has or will request visibleCharacterRange on it.

Removing it saves a substantial amount of memory — 68.1mb on <a href="http://html.spec.whatwg.org">http://html.spec.whatwg.org</a> (567769 objects times 120 bytes).

* LayoutTests/accessibility/visible-character-range-vertical-rl-expected.txt: Added.
* LayoutTests/accessibility/visible-character-range-vertical-rl.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const):
(WebCore::textStartPoint):
(WebCore::textEndPoint):
(WebCore::AccessibilityObject::boundaryPointsContainedInRect const):
(WebCore::AccessibilityObject::visibleCharacterRange const):
(WebCore::AccessibilityObject::visibleCharacterRangeInternal const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const):

Canonical link: <a href="https://commits.webkit.org/286780@main">https://commits.webkit.org/286780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae170613fb4855acb7a00d1d5570a6f30d23597f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77041 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81591 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28320 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79158 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65224 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4372 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60367 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18441 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80108 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50321 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66130 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40678 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47723 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23628 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26646 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68839 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23956 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83025 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4421 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2966 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68650 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4576 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66103 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67903 "Found 1 new API test failure: /TestWebKit:WebKit.InjectedBundleInitializationUserDataCallbackWins (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16942 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11880 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9967 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4367 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7183 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4387 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7822 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6146 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -845,56 +845,72 @@ std::optional<BoundaryPoint> AccessibilityObject::lastBoundaryPointContainedInRe
 
     // Origin isn't contained in visible rect, start moving forward by line.
-    while (!contentRect.contains(elementRect.location())) {
+    while (!contentRect.contains(textStartPoint(elementRect, isFlipped))) {
```
is isFlipped the right word or isVerticalWriting?

twilco replied:
`isFlipped` is accurate. That encapsulates `vertical-rl` and `horizontal-tb` (horizontal top-to-bottom), which both flip the text start and end points (the start is at x + width or y + height rather than at x or y), which is what we care about. We would not want to apply this logic to `vertical-lr`, which is not flipped (doing so would cause the wrong behavior).

https://github.com/WebKit/WebKit/blob/1593e890b56108f2b3ff0cb7b15739ea0db8e624/Source/WebCore/platform/text/WritingMode.h#L136

twilco committed:
AX: AccessibilityObject::visibleCharacterRange() returns the wrong result for vertical writing mode text that is only partially visible
https://bugs.webkit.org/show_bug.cgi?id=283330
rdar://140158374

Reviewed by Chris Fleizach.

AccessibilityObject::visibleCharacterRange() is implemented by shrinking the start and end of the element's rect
line-by-line until it fits within the viewport. The problem was that the logic for determining whether to start or
continue shrinking lines was based on checking rect.location() for shrinking the start of the text, or rect.location() + rect.size()
for the shrinking the end. This is wrong for "flipped" writing-modes (e.g. vertical-rl), whose `x` represents the end
of the text, not the start. This meant we would shrink the end lines when we actually should've been shrinking the start,
and vice versa (hence this problem being specific to "flipped" writing modes).

With this commit, this logic now properly checks for and handles flipped writing modes. A test is added to prove we
behave correctly.

This commit also critically removes AccessibilityObject::m_cachedVisibleCharacterRange and AccessibilityObject::m_cachedVisibleCharacterRangeInputs.
These caches were added to make repeated calls for the same visibleCharacterRange fast, as Books seemed to be doing this
in the past. However, it doesn't seem to be doing this anymore. This cache was not a great idea in the first place (I added it).
It adds 120 bytes of size to every single AccessibilityObject, even if nothing has or will request visibleCharacterRange on it.

Removing it saves a substantial amount of memory — 68.1mb on http://html.spec.whatwg.org (567769 objects times 120 bytes).

* LayoutTests/accessibility/visible-character-range-vertical-rl-expected.txt: Added.
* LayoutTests/accessibility/visible-character-range-vertical-rl.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const):
(WebCore::textStartPoint):
(WebCore::textEndPoint):
(WebCore::AccessibilityObject::boundaryPointsContainedInRect const):
(WebCore::AccessibilityObject::visibleCharacterRange const):
(WebCore::AccessibilityObject::visibleCharacterRangeInternal const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::lastBoundaryPointContainedInRect const):

Canonical link: https://commits.webkit.org/286780@main
